diff -c 'drivers/hid/hid-core.c.OG' 'drivers/hid/hid-core.c'
Index: ./drivers/hid/hid-core.c.OG
*** ./drivers/hid/hid-core.c.OG	Wed Feb 22 21:18:54 2017
--- ./drivers/hid/hid-core.c	Wed Feb 22 21:20:34 2017
***************
*** 2490,2496 ****
--- 2490,2497 ----
  	{ HID_USB_DEVICE(USB_VENDOR_ID_SYNAPTICS, USB_DEVICE_ID_SYNAPTICS_WTP) },
  	{ HID_USB_DEVICE(USB_VENDOR_ID_SYNAPTICS, USB_DEVICE_ID_SYNAPTICS_DPAD) },
  #endif
+ 	{ HID_USB_DEVICE(USB_VENDOR_ID_VERNIER, USB_DEVICE_ID_VERNIER_GOTEMP) },
  	{ HID_USB_DEVICE(USB_VENDOR_ID_YEALINK, USB_DEVICE_ID_YEALINK_P1K_P4K_B2K) },
  	{ }
  };
