diff -Naur iw-3.17/debian/changelog iw-3.17-16ton/debian/changelog
--- iw-3.17/debian/changelog	2014-10-14 14:43:15.000000000 -0400
+++ iw-3.17-16ton/debian/changelog	2018-05-26 00:18:24.518371191 -0400
@@ -1,3 +1,10 @@
+iw (3.17-16ton) unstable; urgency=medium
+
+  * Allow negative channels
+  * Channel to frequency correction
+
+ -- Ben Maddocks <bm16ton@gmail.com>  Sat, 07 Oct 2017 20:42:02 +0200
+
 iw (3.17-1) unstable; urgency=medium
 
   * New upstream release.
diff -Naur iw-3.17/Makefile iw-3.17-16ton/Makefile
--- iw-3.17/Makefile	2018-05-26 00:19:39.000000000 -0400
+++ iw-3.17-16ton/Makefile	2018-01-30 10:20:14.859823808 -0500
@@ -20,7 +20,7 @@
 	interface.o ibss.o station.o survey.o util.o \
 	mesh.o mpath.o scan.o reg.o version.o \
 	reason.o status.o connect.o link.o offch.o ps.o cqm.o \
-	bitrate.o wowlan.o coalesce.o roc.o p2p.o
+	bitrate.o wowlan.o coalesce.o roc.o p2p.o vendor.o
 OBJS += sections.o
 
 OBJS-$(HWSIM) += hwsim.o
diff -Naur iw-3.17/phy.c iw-3.17-16ton/phy.c
--- iw-3.17/phy.c	2014-10-09 10:31:00.000000000 -0400
+++ iw-3.17-16ton/phy.c	2017-10-07 10:21:20.978836611 -0400
@@ -101,6 +101,7 @@
 		{ .name = "HT40-", .val = NL80211_CHAN_HT40MINUS, },
 	};
 	unsigned int htval = NL80211_CHAN_NO_HT;
+	int val;
 	unsigned int freq;
 	int i;
 
@@ -109,14 +110,19 @@
 
 	if (!*argv[0])
 		return 1;
-	freq = strtoul(argv[0], &end, 10);
+	val = strtol(argv[0], &end, 10);
 	if (*end)
 		return 1;
 
+   // treat val as channel number
 	if (chan) {
 		enum nl80211_band band;
-		band = freq <= 14 ? NL80211_BAND_2GHZ : NL80211_BAND_5GHZ;
-		freq = ieee80211_channel_to_frequency(freq, band);
+		band = val <= 14 ? NL80211_BAND_2GHZ : NL80211_BAND_5GHZ;
+		freq = ieee80211_channel_to_frequency(val, band);
+	}
+	else // treat val as frequency and cut a possible sign 
+	{
+		freq = (unsigned long) val;
 	}
 
 	NLA_PUT_U32(msg, NL80211_ATTR_WIPHY_FREQ, freq);
diff -Naur iw-3.17/util.c iw-3.17-16ton/util.c
--- iw-3.17/util.c	2014-10-09 10:31:00.000000000 -0400
+++ iw-3.17-16ton/util.c	2017-10-08 09:48:29.158067971 -0400
@@ -268,14 +268,16 @@
 {
 	/* see 802.11 17.3.8.3.2 and Annex J
 	 * there are overlapping channel numbers in 5GHz and 2GHz bands */
-	if (chan <= 0)
-		return 0; /* not supported */
+	//if (chan <= 0)
+	//	return 0; /* not supported */
 	switch (band) {
 	case NL80211_BAND_2GHZ:
 		if (chan == 14)
 			return 2484;
 		else if (chan < 14)
 			return 2407 + chan * 5;
+		else if (chan > 14 && chan <= 65)
+			return 2407 + chan * 5;
 		break;
 	case NL80211_BAND_5GHZ:
 		if (chan >= 182 && chan <= 196)
@@ -300,6 +302,8 @@
 		return 14;
 	else if (freq < 2484)
 		return (freq - 2407) / 5;
+	else if (freq > 2484 && freq <= 2732)
+		return (freq - 2407) / 5;
 	else if (freq >= 4910 && freq <= 4980)
 		return (freq - 4000) / 5;
 	else if (freq <= 45000) /* DMG band lower limit */
diff -Naur iw-3.17/vendor.c iw-3.17-16ton/vendor.c
--- iw-3.17/vendor.c	1969-12-31 19:00:00.000000000 -0500
+++ iw-3.17-16ton/vendor.c	2018-01-30 10:15:02.923226401 -0500
@@ -0,0 +1,95 @@
+#include <net/if.h>
+#include <errno.h>
+#include <string.h>
+
+#include <netlink/genl/genl.h>
+#include <netlink/genl/family.h>
+#include <netlink/genl/ctrl.h>
+#include <netlink/msg.h>
+#include <netlink/attr.h>
+
+#include "nl80211.h"
+#include "iw.h"
+
+SECTION(vendor);
+
+static int read_file(FILE *file, char *buf, size_t size)
+{
+	int data, count = 0;
+
+	while ((data = fgetc(file)) != EOF) {
+		if (count >= size)
+			return -EINVAL;
+		buf[count] = data;
+		count++;
+	}
+
+	return count;
+}
+
+static int read_hex(int argc, char **argv, char *buf, size_t size)
+{
+	int i, res;
+	unsigned int data;
+
+	if (argc > size)
+		return -EINVAL;
+
+	for (i = 0; i < argc; i++) {
+		res = sscanf(argv[i], "0x%x", &data);
+		if (res != 1 || data > 0xff)
+			return -EINVAL;
+		buf[i] = data;
+	}
+
+	return argc;
+}
+
+static int handle_vendor(struct nl80211_state *state, struct nl_cb *cb,
+			 struct nl_msg *msg, int argc, char **argv,
+			 enum id_input id)
+{
+	unsigned int oui;
+	unsigned int subcmd;
+	char buf[2048] = {};
+	int res, count = 0;
+	FILE *file = NULL;
+
+	if (argc < 3)
+		return -EINVAL;
+
+	res = sscanf(argv[0], "0x%x", &oui);
+	if (res != 1)
+		return -EINVAL;
+
+	res = sscanf(argv[1], "0x%x", &subcmd);
+	if (res != 1)
+		return -EINVAL;
+
+	if (!strcmp(argv[2], "-"))
+		file = stdin;
+	else
+		file = fopen(argv[2], "r");
+
+	NLA_PUT_U32(msg, NL80211_ATTR_VENDOR_ID, oui);
+	NLA_PUT_U32(msg, NL80211_ATTR_VENDOR_SUBCMD, subcmd);
+
+	if (file) {
+		count = read_file(file, buf, sizeof(buf));
+		fclose(file);
+	} else
+		count = read_hex(argc - 2, &argv[2], buf, sizeof(buf));
+
+	if (count < 0)
+		return -EINVAL;
+
+	if (count > 0)
+		NLA_PUT(msg, NL80211_ATTR_VENDOR_DATA, count, buf);
+
+	return 0;
+
+nla_put_failure:
+	return -ENOBUFS;
+}
+
+COMMAND(vendor, send, "<oui> <subcmd> <filename|-|hex data>", NL80211_CMD_VENDOR, 0, CIB_NETDEV, handle_vendor, "");
diff -Naur iw-3.17/version.sh iw-3.17-16ton/version.sh
--- iw-3.17/version.sh	2014-10-09 10:31:00.000000000 -0400
+++ iw-3.17-16ton/version.sh	2017-10-07 11:05:06.172461939 -0400
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-VERSION="3.17"
+VERSION="3.17-16ton"
 OUT="$1"
 
 if [ -d .git ] && head=`git rev-parse --verify HEAD 2>/dev/null`; then
