diff -Naur iw-3.17/debian/changelog iw-3.17-16ton/debian/changelog
--- iw-3.17/debian/changelog	2014-10-14 14:43:15.000000000 -0400
+++ iw-3.17-16ton/debian/changelog	2017-10-10 14:06:07.563617899 -0400
@@ -1,3 +1,10 @@
+iw (3.17-16ton) unstable; urgency=medium
+
+  * Allow negative channels
+  * Channel to frequency correction
+
+ -- Ben Maddocks <bm16ton@gmail.com>  Sat, 07 Oct 2017 20:42:02 +0200
+
 iw (3.17-1) unstable; urgency=medium
 
   * New upstream release.
diff -Naur iw-3.17/phy.c iw-3.17-16ton/phy.c
--- iw-3.17/phy.c	2014-10-09 10:31:00.000000000 -0400
+++ iw-3.17-16ton/phy.c	2017-10-07 10:21:20.978836611 -0400
@@ -101,6 +101,7 @@
 		{ .name = "HT40-", .val = NL80211_CHAN_HT40MINUS, },
 	};
 	unsigned int htval = NL80211_CHAN_NO_HT;
+	int val;
 	unsigned int freq;
 	int i;
 
@@ -109,14 +110,19 @@
 
 	if (!*argv[0])
 		return 1;
-	freq = strtoul(argv[0], &end, 10);
+	val = strtol(argv[0], &end, 10);
 	if (*end)
 		return 1;
 
+   // treat val as channel number
 	if (chan) {
 		enum nl80211_band band;
-		band = freq <= 14 ? NL80211_BAND_2GHZ : NL80211_BAND_5GHZ;
-		freq = ieee80211_channel_to_frequency(freq, band);
+		band = val <= 14 ? NL80211_BAND_2GHZ : NL80211_BAND_5GHZ;
+		freq = ieee80211_channel_to_frequency(val, band);
+	}
+	else // treat val as frequency and cut a possible sign 
+	{
+		freq = (unsigned long) val;
 	}
 
 	NLA_PUT_U32(msg, NL80211_ATTR_WIPHY_FREQ, freq);
diff -Naur iw-3.17/version.sh iw-3.17-16ton/version.sh
--- iw-3.17/version.sh	2014-10-09 10:31:00.000000000 -0400
+++ iw-3.17-16ton/version.sh	2017-10-07 11:05:06.172461939 -0400
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-VERSION="3.17"
+VERSION="3.17-16ton"
 OUT="$1"
 
 if [ -d .git ] && head=`git rev-parse --verify HEAD 2>/dev/null`; then
