diff --git a/drivers/nfc/nfcmrvl/usb.c b/drivers/nfc/nfcmrvl/usb.c
--- a/drivers/nfc/nfcmrvl/usb.c
+++ b/drivers/nfc/nfcmrvl/usb.c
@@ -26,8 +26,7 @@
 #define VERSION "1.0"

 static struct usb_device_id nfcmrvl_table[] = {
-	{ USB_DEVICE_AND_INTERFACE_INFO(0x1286, 0x2046,
-					USB_CLASS_VENDOR_SPEC, 4, 1) },
+	{ USB_DEVICE_INTERFACE_CLASS(0x1286, 0x2046, 0xff) },
 	{ }	/* Terminating entry */
 };

