--- linux-4.8.0-41.44/drivers/net/wireless/marvell/mwifiex/Kconfig	2016-10-02 19:24:33.000000000 -0400
+++ linux-hwe-4.8.0/drivers/net/wireless/marvell/mwifiex/Kconfig	2017-09-03 09:50:17.353989268 -0400
@@ -33,12 +33,12 @@
 	  mwifiex_pcie.
 
 config MWIFIEX_USB
-	tristate "Marvell WiFi-Ex Driver for USB8766/8797/8997"
+	tristate "Marvell WiFi-Ex Driver for USB8766/8797/8897/8997"
 	depends on MWIFIEX && USB
 	select FW_LOADER
 	---help---
 	  This adds support for wireless adapters based on Marvell
-	  8797/8997 chipset with USB interface.
+	  8797/8897/8997 chipset with USB interface.
 
 	  If you choose to build it as a module, it will be called
 	  mwifiex_usb.
