--- drivers/net/wireless/ath.OG/ath9k/init.c	2017-10-01 20:33:30.000000000 -0400
+++ drivers/net/wireless/ath/ath9k/init.c	2017-10-02 15:43:24.546912802 -0400
@@ -73,6 +73,12 @@
 
 #endif /* CONFIG_ATH9K_CHANNEL_CONTEXT */
 
+static int modparam_override_eeprom_regdomain = -1;
+module_param_named(override_eeprom_regdomain,
+ modparam_override_eeprom_regdomain, int, 0444);
+MODULE_PARM_DESC(override_eeprom_regdomain, "Override regdomain hardcoded in EEPROM with this value (DANGEROUS).");
+
+
 bool is_ath9k_unloaded;
 
 #ifdef CONFIG_MAC80211_LEDS
@@ -814,6 +820,21 @@
 	struct ath_hw *ah = sc->sc_ah;
 	struct ath_common *common = ath9k_hw_common(ah);
 
+	if (modparam_override_eeprom_regdomain != -1) {
+#ifdef CONFIG_CFG80211_CERTIFICATION_ONUS
+	struct ath_regulatory *regulatory = ath9k_hw_regulatory(sc->sc_ah);
+	printk(KERN_ERR "ath9k: DANGER! You're overriding EEPROM-defined regulatory domain,"
+	"\nfrom: 0x%x to 0x%x\n",
+	regulatory->current_rd, modparam_override_eeprom_regdomain);
+	printk(KERN_ERR "ath9k: Your card was not certified to operate in the domain you chose.\n");
+	printk(KERN_ERR "ath9k: This might result in a violation of your local regulatory rules.\n");
+	printk(KERN_ERR "ath9k: Do not ever do this unless you really know what you are doing!\n");
+	regulatory->current_rd = modparam_override_eeprom_regdomain;
+#else
+	printk(KERN_ERR "ath9k: ERROR: override_eeprom_regdomain request will be ignored because CF80211_CERTIFICATION_ONUS was not selected when compiling the kernel.\n");
+#endif
+}
+
 	ieee80211_hw_set(hw, SUPPORTS_HT_CCK_RATES);
 	ieee80211_hw_set(hw, SUPPORTS_RC_TABLE);
 	ieee80211_hw_set(hw, REPORTS_TX_ACK_STATUS);
