
diff -upr hostapd-2.5.orig/src/ap/hw_features.c hostapd-2.5/src/ap/hw_features.c
--- hostapd-2.5.orig/src/ap/hw_features.c       2015-09-27 21:02:05.000000000 +0200
+++ hostapd-2.5/src/ap/hw_features.c    2016-03-15 17:00:00.163082949 +0100
@@ -316,7 +316,7 @@ static void ieee80211n_check_scan(struct
                           iface->conf->channel,
                           iface->conf->channel +
                           iface->conf->secondary_channel * 4);
-               iface->conf->secondary_channel = 0;
+               /* iface->conf->secondary_channel = 0; DILLIGAF */
                if (iface->drv_flags & WPA_DRIVER_FLAGS_HT_2040_COEX) {
                        /*
                         * TODO: Could consider scheduling another scan to check
diff -upr hostapd-2.5.orig/src/ap/ieee802_11_ht.c hostapd-2.5/src/ap/ieee802_11_ht.c
--- hostapd-2.5.orig/src/ap/ieee802_11_ht.c     2015-09-27 21:02:05.000000000 +0200
+++ hostapd-2.5/src/ap/ieee802_11_ht.c  2016-03-15 17:04:06.571590399 +0100
@@ -282,6 +282,7 @@ void hostapd_2040_coex_action(struct hos
        wpa_printf(MSG_DEBUG, "is_ht40_allowed=%d num_sta_ht40_intolerant=%d",
                   is_ht40_allowed, iface->num_sta_ht40_intolerant);
 
+    /* DILLIGAF
        if (!is_ht40_allowed &&
            (iface->drv_flags & WPA_DRIVER_FLAGS_HT_2040_COEX)) {
                if (iface->conf->secondary_channel) {
@@ -306,6 +307,7 @@ void hostapd_2040_coex_action(struct hos
                                   delay_time);
                }
        }
+    */
 }
