diff -Naur hostapd-2.2.orig/src/ap/hw_features.c hostapd-2.2/src/ap/hw_features.c
--- hostapd-2.2.orig/src/ap/hw_features.c       2014-06-04 15:26:14.000000000 +0200
+++ hostapd-2.2/src/ap/hw_features.c    2014-06-16 14:58:26.740289954 +0200
@@ -539,7 +539,7 @@
                           iface->conf->channel,
                           iface->conf->channel +
                           iface->conf->secondary_channel * 4);
-               iface->conf->secondary_channel = 0;
+               /* iface->conf->secondary_channel = 0; DILLIGAF */
                if (iface->drv_flags & WPA_DRIVER_FLAGS_HT_2040_COEX) {
                        /*
                         * TODO: Could consider scheduling another scan to check
