
--- hostapd-2.6.orig/src/ap/hw_features.c       2016-10-02 20:51:11.000000000 +0200
+++ hostapd-2.6/src/ap/hw_features.c    2016-10-11 12:11:28.430817177 +0200
@@ -316,7 +316,7 @@
                           iface->conf->channel,
                           iface->conf->channel +
                           iface->conf->secondary_channel * 4);
-               iface->conf->secondary_channel = 0;
+               /* iface->conf->secondary_channel = 0; DILLIGAF */
                if (iface->drv_flags & WPA_DRIVER_FLAGS_HT_2040_COEX) {
                        /*
                         * TODO: Could consider scheduling another scan to check
--- hostapd-2.6.orig/src/ap/ieee802_11_ht.c     2016-10-02 20:51:11.000000000 +0200
+++ hostapd-2.6/src/ap/ieee802_11_ht.c  2016-10-11 12:14:18.473952410 +0200
@@ -305,6 +305,7 @@
        wpa_printf(MSG_DEBUG, "is_ht40_allowed=%d num_sta_ht40_intolerant=%d",
                   is_ht40_allowed, iface->num_sta_ht40_intolerant);
 
+/* DILLIGAF
        if (!is_ht40_allowed &&
            (iface->drv_flags & WPA_DRIVER_FLAGS_HT_2040_COEX)) {
                if (iface->conf->secondary_channel) {
@@ -329,6 +330,7 @@
                                   delay_time);
                }
        }
+*/
 }
