diff -Naur linux-4.14.24/scripts/setlocalversion linux-gpd-4.14.24/scripts/setlocalversion
--- linux-4.14.24/scripts/setlocalversion	2018-03-03 04:24:39.000000000 -0500
+++ linux-gpd-4.14.24/scripts/setlocalversion	2018-03-06 18:02:50.105286034 -0500
@@ -161,15 +161,15 @@
 if test "$CONFIG_LOCALVERSION_AUTO" = "y"; then
 	# full scm version string
 	res="$res$(scm_version)"
-else
+#else
 	# append a plus sign if the repository is not in a clean
 	# annotated or signed tagged state (as git describe only
 	# looks at signed or annotated tags - git tag -a/-s) and
 	# LOCALVERSION= is not specified
-	if test "${LOCALVERSION+set}" != "set"; then
-		scm=$(scm_version --short)
-		res="$res${scm:++}"
-	fi
+#	if test "${LOCALVERSION+set}" != "set"; then
+#		scm=$(scm_version --short)
+#		res="$res${scm:++}"
+#	fi
 fi
 
 echo "$res"
