diff -Naur linux-4.8/fs/open.c zen-kernel-4.8-misc/fs/open.c
--- linux-4.8/fs/open.c	2016-10-02 19:24:33.000000000 -0400
+++ zen-kernel-4.8-misc/fs/open.c	2016-12-04 16:58:54.000000000 -0500
@@ -34,6 +34,9 @@
 
 #include "internal.h"
 
+#define CREATE_TRACE_POINTS
+#include <trace/events/fs.h>
+
 int do_truncate(struct dentry *dentry, loff_t length, unsigned int time_attrs,
 	struct file *filp)
 {
@@ -1040,6 +1043,7 @@
 		} else {
 			fsnotify_open(f);
 			fd_install(fd, f);
+			trace_do_sys_open(tmp->name, flags, mode);
 		}
 	}
 	putname(tmp);
