diff -Naur linux-hwe-4.8.0/drivers/net/wireless/realtek/rtlwifi/base.c linux-wifibrdcst-4.8.17/drivers/net/wireless/realtek/rtlwifi/base.c
--- linux-hwe-4.8.0/drivers/net/wireless/realtek/rtlwifi/base.c	2017-10-10 11:01:30.000000000 -0400
+++ linux-wifibrdcst-4.8.17/drivers/net/wireless/realtek/rtlwifi/base.c	2017-10-13 08:38:21.357020860 -0400
@@ -61,6 +61,26 @@
  *
  *********************************************************/
 static struct ieee80211_channel rtl_channeltable_2g[] = {
+	{.center_freq = 2312, .hw_value = -19,},
+	{.center_freq = 2317, .hw_value = -18,},
+	{.center_freq = 2322, .hw_value = -17,},
+	{.center_freq = 2327, .hw_value = -16,},
+	{.center_freq = 2332, .hw_value = -15,},
+	{.center_freq = 2337, .hw_value = -14,},
+	{.center_freq = 2342, .hw_value = -13,},
+	{.center_freq = 2347, .hw_value = -12,},
+	{.center_freq = 2352, .hw_value = -11,},
+	{.center_freq = 2357, .hw_value = -10,},
+	{.center_freq = 2362, .hw_value = -9,},
+	{.center_freq = 2367, .hw_value = -8,},
+	{.center_freq = 2372, .hw_value = -7,},
+	{.center_freq = 2377, .hw_value = -6,},
+	{.center_freq = 2382, .hw_value = -5,},
+	{.center_freq = 2387, .hw_value = -4,},
+	{.center_freq = 2392, .hw_value = -3,},
+	{.center_freq = 2397, .hw_value = -2,},
+	{.center_freq = 2402, .hw_value = -1,},
+	{.center_freq = 2407, .hw_value = 0,},
 	{.center_freq = 2412, .hw_value = 1,},
 	{.center_freq = 2417, .hw_value = 2,},
 	{.center_freq = 2422, .hw_value = 3,},
@@ -75,6 +95,18 @@
 	{.center_freq = 2467, .hw_value = 12,},
 	{.center_freq = 2472, .hw_value = 13,},
 	{.center_freq = 2484, .hw_value = 14,},
+	{.center_freq = 2487, .hw_value = 16,},
+	{.center_freq = 2502, .hw_value = 19,},
+	{.center_freq = 2512, .hw_value = 21,},
+	{.center_freq = 2572, .hw_value = 33,},
+	{.center_freq = 2592, .hw_value = 37,},
+	{.center_freq = 2612, .hw_value = 41,},
+	{.center_freq = 2632, .hw_value = 45,},
+	{.center_freq = 2652, .hw_value = 49,},
+	{.center_freq = 2672, .hw_value = 53,},
+	{.center_freq = 2692, .hw_value = 57,},
+	{.center_freq = 2712, .hw_value = 61,},
+	{.center_freq = 2732, .hw_value = 65,},
 };
 
 static struct ieee80211_channel rtl_channeltable_5g[] = {
diff -Naur linux-hwe-4.8.0/drivers/net/wireless/realtek/rtlwifi/regd.c linux-wifibrdcst-4.8.17/drivers/net/wireless/realtek/rtlwifi/regd.c
--- linux-hwe-4.8.0/drivers/net/wireless/realtek/rtlwifi/regd.c	2017-10-10 11:01:30.000000000 -0400
+++ linux-wifibrdcst-4.8.17/drivers/net/wireless/realtek/rtlwifi/regd.c	2017-10-13 08:39:21.165965075 -0400
@@ -48,25 +48,22 @@
  *scan on all world regulatory domains
  */
 #define RTL819x_2GHZ_CH01_11	\
-	REG_RULE(2412-10, 2462+10, 40, 0, 20, 0)
+	REG_RULE(2312-10, 2462+10, 40, 0, 20, 0)
 
 /*
  *We enable active scan on these a case
  *by case basis by regulatory domain
  */
 #define RTL819x_2GHZ_CH12_13	\
-	REG_RULE(2467-10, 2472+10, 40, 0, 20,\
-	NL80211_RRF_PASSIVE_SCAN)
+	REG_RULE(2467-10, 2472+10, 40, 0, 20, 0)
 
 #define RTL819x_2GHZ_CH14	\
-	REG_RULE(2484-10, 2484+10, 40, 0, 20, \
-	NL80211_RRF_PASSIVE_SCAN | \
-	NL80211_RRF_NO_OFDM)
+	REG_RULE(2484-10, 2732+10, 40, 0, 20, 0)
 
 
 /* 5G chan 36 - chan 64*/
 #define RTL819x_5GHZ_5150_5350	\
-	REG_RULE(5150-10, 5350+10, 80, 0, 30, 0)
+	REG_RULE(4920-10, 5350+10, 80, 0, 30, 0)
 /* 5G chan 100 - chan 165*/
 #define RTL819x_5GHZ_5470_5850	\
 	REG_RULE(5470-10, 5850+10, 80, 0, 30, 0)
@@ -82,6 +79,8 @@
 	.alpha2 = "99",
 	.reg_rules = {
 		      RTL819x_2GHZ_CH01_11,
+			  RTL819x_2GHZ_CH12_13,
+			  RTL819x_2GHZ_CH14,
 		      }
 };
 
@@ -91,6 +90,7 @@
 	.reg_rules = {
 		      RTL819x_2GHZ_CH01_11,
 			  RTL819x_2GHZ_CH12_13,
+			  RTL819x_2GHZ_CH14,
 		      }
 };
 
@@ -99,8 +99,10 @@
 	.alpha2 = "99",
 	.reg_rules = {
 		      RTL819x_2GHZ_CH01_11,
+			  RTL819x_2GHZ_CH12_13,
+			  RTL819x_2GHZ_CH14,
 			  RTL819x_5GHZ_5150_5350,
-			  RTL819x_5GHZ_5725_5850,
+	          RTL819x_5GHZ_5470_5850,
 		      }
 };
 
@@ -110,7 +112,9 @@
 	.reg_rules = {
 		      RTL819x_2GHZ_CH01_11,
 			  RTL819x_2GHZ_CH12_13,
-			  RTL819x_5GHZ_5725_5850,
+			  RTL819x_2GHZ_CH14,
+			  RTL819x_5GHZ_5150_5350,
+	          RTL819x_5GHZ_5470_5850,
 		      }
 };
 
@@ -121,7 +125,8 @@
 		      RTL819x_2GHZ_CH01_11,
 			  RTL819x_2GHZ_CH12_13,
 			  RTL819x_2GHZ_CH14,
-			  RTL819x_5GHZ_5725_5850,
+			  RTL819x_5GHZ_5150_5350,
+	          RTL819x_5GHZ_5470_5850,
 		      }
 };
 
@@ -129,8 +134,9 @@
 	.n_reg_rules = 4,
 	.alpha2 = "99",
 	.reg_rules = {
-			RTL819x_2GHZ_CH01_11,
-			RTL819x_2GHZ_CH12_13,
+		      RTL819x_2GHZ_CH01_11,
+			  RTL819x_2GHZ_CH12_13,
+			  RTL819x_2GHZ_CH14,
 			RTL819x_5GHZ_5150_5350,
 			RTL819x_5GHZ_5470_5850,
 		}
@@ -148,7 +154,8 @@
 
 static bool _rtl_is_radar_freq(u16 center_freq)
 {
-	return center_freq >= 5260 && center_freq <= 5700;
+//	return center_freq >= 5260 && center_freq <= 5700;
+	return 0;
 }
 
 static void _rtl_reg_apply_beaconing_flags(struct wiphy *wiphy,
