diff -urpN ubuntu-mtp-71_orig/src/MtpServer.cpp ubuntu-mtp-71/src/MtpServer.cpp
--- ubuntu-mtp-71_orig/src/MtpServer.cpp	2016-04-11 03:32:19.000000000 -0500
+++ ubuntu-mtp-71/src/MtpServer.cpp	2016-11-16 09:30:48.323260965 -0600
@@ -41,7 +41,6 @@
 
 #include <linux/usb/f_mtp.h>
 
-#include <hybris/properties/properties.h>
 
 #include <glog/logging.h>
 
@@ -468,7 +467,7 @@ bool MtpServer::handleRequest() {
 MtpResponseCode MtpServer::doGetDeviceInfo() {
     VLOG(1) <<  __PRETTY_FUNCTION__;
     MtpStringBuffer   string;
-    char prop_value[PROP_VALUE_MAX];
+    char prop_value[64];
 
     MtpObjectFormatList* playbackFormats = mDatabase->getSupportedPlaybackFormats();
     MtpObjectFormatList* captureFormats = mDatabase->getSupportedCaptureFormats();
@@ -500,17 +499,17 @@ MtpResponseCode MtpServer::doGetDeviceIn
     mData.putAUInt16(captureFormats); // Capture Formats
     mData.putAUInt16(playbackFormats);  // Playback Formats
 
-    property_get("ro.product.manufacturer", prop_value, "unknown manufacturer");
+    strcpy(prop_value,"---MANUFACTURER---");
     string.set(prop_value);
     mData.putString(string);   // Manufacturer
 
-    property_get("ro.product.model", prop_value, "MTP Device");
+    strcpy(prop_value,"---MODEL---");
     string.set(prop_value);
     mData.putString(string);   // Model
     string.set("1.0");
     mData.putString(string);   // Device Version
 
-    property_get("ro.serialno", prop_value, "????????");
+    strcpy(prop_value,"---SERIAL---");
     string.set(prop_value);
     mData.putString(string);   // Serial Number
 
