--- drivers/net/wireless/ath.OG/ath9k/eeprom.h	2017-10-02 20:19:47.915512493 -0400
+++ drivers/net/wireless/ath/ath9k/eeprom.h	2017-10-02 20:18:13.452162997 -0400
@@ -642,6 +642,7 @@
 struct eeprom_ops {
 	int (*check_eeprom)(struct ath_hw *hw);
 	u32 (*get_eeprom)(struct ath_hw *hw, enum eeprom_param param);
+	u32 (*dump_eep_power)(struct ath_hw *hw, u8 *buf, u32 len, u32 size);
 	bool (*fill_eeprom)(struct ath_hw *hw);
 	u32 (*dump_eeprom)(struct ath_hw *hw, bool dump_base_hdr, u8 *buf,
 			   u32 len, u32 size);
