--- a/phy.c
+++ b/phy.c
@@ -101,6 +101,7 @@
 		{ .name = "HT40-", .val = NL80211_CHAN_HT40MINUS, },
 	};
 	unsigned int htval = NL80211_CHAN_NO_HT;
+	int val;
 	unsigned int freq;
 	int i;
 
@@ -109,14 +110,19 @@
 
 	if (!*argv[0])
 		return 1;
-	freq = strtoul(argv[0], &end, 10);
+	val = strtol(argv[0], &end, 10);
 	if (*end)
 		return 1;
 
+   // treat val as channel number
 	if (chan) {
 		enum nl80211_band band;
-		band = freq <= 14 ? NL80211_BAND_2GHZ : NL80211_BAND_5GHZ;
-		freq = ieee80211_channel_to_frequency(freq, band);
+		band = val <= 14 ? NL80211_BAND_2GHZ : NL80211_BAND_5GHZ;
+		freq = ieee80211_channel_to_frequency(val, band);
+	}
+	else // treat val as frequency and cut a possible sign 
+	{
+		freq = (unsigned long) val;
 	}
 
 	NLA_PUT_U32(msg, NL80211_ATTR_WIPHY_FREQ, freq);
