diff --git a/kernel/sched/bfs.c b/kernel/sched/bfs.c
index 8cf920ebf2a7..fadcf2df1d94 100644
--- a/kernel/sched/bfs.c
+++ b/kernel/sched/bfs.c
@@ -245,7 +245,7 @@ static inline struct rq
 				return rq;
 			}
 			raw_spin_unlock(&rq->lock);
-		} else if (task_queued(p)) {
+		} else if (p->on_rq) {
 			raw_spin_lock(&rq->lock);
 			if (likely(!p->on_cpu && task_queued(p) &&
 				   rq == task_rq(p))) {
@@ -281,9 +281,9 @@ struct rq
 				return rq;
 			}
 			raw_spin_unlock_irqrestore(&rq->lock, *flags);
-		} else if (task_queued(p)) {
+		} else if (p->on_rq) {
 			raw_spin_lock_irqsave(&rq->lock, *flags);
-			if (likely(!p->on_cpu && task_queued(p) &&
+			if (likely(!p->on_cpu && p->on_rq &&
 				   rq == task_rq(p))) {
 				*plock = &rq->lock;
 				return rq;
@@ -291,7 +291,7 @@ struct rq
 			raw_spin_unlock_irqrestore(&rq->lock, *flags);
 		} else {
 			raw_spin_lock_irqsave(&p->pi_lock, *flags);
-			if (likely(!p->on_cpu && !task_queued(p) &&
+			if (likely(!p->on_cpu && !p->on_rq &&
 				   rq == task_rq(p))) {
 				*plock = &p->pi_lock;
 				return rq;
@@ -1741,9 +1741,7 @@ static inline void ttwu_activate(struct task_struct *p, struct rq *rq)
 	activate_task(p, rq);
 
 	/*
-	 * if a worker is waking up, notify workqueue. Note that on BFS, we
-	 * don't really know what CPU it will be, so we fake it for
-	 * wq_worker_waking_up :/
+	 * if a worker is waking up, notify workqueue.
 	 */
 	if (p->flags & PF_WQ_WORKER)
 		wq_worker_waking_up(p, cpu_of(rq));
